[JSC] Improve Performance of `Math.sumPrecise` for Large Inputs

Gumichocopengin8 opened the pull request:
#### d95d6a80355e27cd4091e806e21019e4bc557a1e
<pre>
[JSC] Improve Performance of `Math.sumPrecise` for Large Inputs
<a href="https://bugs.webkit.org/show_bug.cgi?id=293866">https://bugs.webkit.org/show_bug.cgi?id=293866</a>

Reviewed by Yusuke Suzuki.

Math.sumPrecise may be used in scientific calculations, charting, and statistical analysis,
where it is crucial to handle large inputs precisely and efficiently.
This patch improves performance for large inputs by implementing Radford M. Neal&apos;s xsum
algorithm using the large superaccumulator[1].
The large superaccumulator is used when the input is an array and its size exceeds
PRECISE_SUM_THRESHOLD (1,000); otherwise, the small superaccumulator is used.
The original implementation of xsum large superaccumulator is in C [2]; a C++ translation
by Keita Nonaka was referenced for integration [3].
This patch also refactors the previous implementation[4] of xsum using the small
superaccumulator to align with the xsum large superaccumulator design.

                                TipOfTree              Patched                Ratio
math-sumPrecise-10              2.6139+-0.1125         2.6545+-0.1376         0.0152x slower
math-sumPrecise-100             17.9892+-0.0930        18.0620+-0.1747        0.0040x slower
math-sumPrecise-1000            172.3968+-1.8629       155.9089+-1.3315       1.1057x faster
math-sumPrecise-10000           1709.5876+-4.2845      1530.4218+-5.2034      1.1170x faster

[1]: <a href="https://arxiv.org/abs/1505.05571">https://arxiv.org/abs/1505.05571</a>
[2]: <a href="https://gitlab.com/radfordneal/xsum">https://gitlab.com/radfordneal/xsum</a>
[3]: <a href="https://github.com/Gumichocopengin8/xsum.cpp">https://github.com/Gumichocopengin8/xsum.cpp</a>
[4]: <a href="https://github.com/WebKit/WebKit/commit/2682d2ab17a40851f7e6bcff35126e810b0646fd">https://github.com/WebKit/WebKit/commit/2682d2ab17a40851f7e6bcff35126e810b0646fd</a>

* JSTests/stress/math-sum-precise.js:
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/PreciseSum.cpp:
(WTF::Xsum::SmallAccumulator::SmallAccumulator):
(WTF::Xsum::SmallAccumulator::addInfNan):
(WTF::Xsum::SmallAccumulator::carryPropagate):
(WTF::Xsum::SmallAccumulator::add1NoCarry):
(WTF::Xsum::SmallAccumulator::incrementWhenValueAdded):
(WTF::Xsum::LargeAccumulator::LargeAccumulator):
(WTF::Xsum::LargeAccumulator::addLchunkToSmall):
(WTF::Xsum::LargeAccumulator::largeAddValueInfNan):
(WTF::Xsum::LargeAccumulator::transferToSmall):
(WTF::Xsum::XsumSmall::XsumSmall):
(WTF::Xsum::XsumSmall::addList):
(WTF::Xsum::XsumSmall::add):
(WTF::Xsum::XsumSmall::compute):
(WTF::Xsum::XsumLarge::XsumLarge):
(WTF::Xsum::XsumLarge::addList):
(WTF::Xsum::XsumLarge::add):
(WTF::Xsum::XsumLarge::compute):
(WTF::PreciseSum::PreciseSum): Deleted.
(WTF::PreciseSum::addList): Deleted.
(WTF::PreciseSum::add): Deleted.
(WTF::PreciseSum::compute): Deleted.
(WTF::PreciseSum::xsumSmallAddInfNan): Deleted.
(WTF::PreciseSum::xsumAdd1NoCarry): Deleted.
(WTF::PreciseSum::xsumCarryPropagate): Deleted.
(WTF::PreciseSum::incrementWhenValueAdded): Deleted.
* Source/WTF/wtf/PreciseSum.h:

Canonical link: <a href="https://commits.webkit.org/296157@main">https://commits.webkit.org/296157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f5546f2af7684fad49e40813f8a7d5cfa9e62fe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107528 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27210 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112744 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58067 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35712 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/81633 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110457 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22109 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96917 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62014 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21548 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15055 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57508 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100120 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91478 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115844 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106077 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34595 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25510 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90670 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34972 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93167 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90411 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23059 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35331 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13107 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30356 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34516 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/40073 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/130391 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34263 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/35457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37622 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35926 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -450,7 +450,10 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
 
-    PreciseSum sum;
+    uint64_t length = toLength(globalObject, iterable.toObject(globalObject));
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
`toLength`, `toObject` etc. are side-effectful. Thus, you cannot do that here. (if an iterable is not array etc., then it performs `get-by-id` for `"length"`, and it is not defined in the spec, and user can observe this behavior by setting a getter for this field).

Gumichocopengin8 replied:
Does that mean there's no way to get the `length` info for the iterable before `forEachInIterable`?
If not, it is probably difficult to optimize. Regarding to the microbenchmark, `enum` + `union` approach requires if branches and it make the function significantly slow.

Gumichocopengin8 replied:
Ended up using a template instead of `union` and `enum` because `switch` or `if` statement for `enum` branching makes the code slow.
Additionally, use `XsumLarge` only if `iterable` is an array type, otherwise fall back to `XsumSmall` as before

Constellation commented on Source/WTF/wtf/PreciseSum.h:
```diff
@@ -32,35 +32,91 @@ namespace WTF {
-    inline void xsumAdd1NoCarry(double value);
-    int xsumCarryPropagate();
-    ALWAYS_INLINE void incrementWhenValueAdded(double value);
+    std::unique_ptr<Xsum::XsumInterface> m_accumulator;
```
I don't feel like this is great for perf as it is always including heap allocation & frequent virtual function calls. Given that there is only two modes, switching them via enum would be better for perf. (And in particular, because this patch is for perf, we should do it).

Constellation commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -450,20 +450,33 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
 
-    PreciseSum sum;
+    std::optional<uint64_t> length = std::nullopt;
+    if (JSC::isArray(globalObject, iterable)) {
```
Use `isJSArray` instead. `isArray` can handle Proxy etc. too.

Gumichocopengin8 replied:
Fixed

Constellation commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -450,20 +450,33 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
-    PreciseSum sum;
+    std::optional<uint64_t> length = std::nullopt;
+    if (JSC::isArray(globalObject, iterable)) {
+        length = toLength(globalObject, iterable.toObject(globalObject));
```
If we find it is `JSArray`, you should use `jsCast<JSArray*>(iterable)` instead of `toObject`.

Constellation replied:
So, we should use `jsCast<JSArray*>(iterable)->length()` instead of `toLength`.

Constellation commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -450,20 +450,33 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
+    std::optional<uint64_t> length = std::nullopt;
+    if (JSC::isArray(globalObject, iterable)) {
+        length = toLength(globalObject, iterable.toObject(globalObject));
+        RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
Let's remove this.

Gumichocopengin8 committed:
[JSC] Improve Performance of `Math.sumPrecise` for Large Inputs
https://bugs.webkit.org/show_bug.cgi?id=293866

Reviewed by Yusuke Suzuki.

Math.sumPrecise may be used in scientific calculations, charting, and statistical analysis,
where it is crucial to handle large inputs precisely and efficiently.
This patch improves performance for large inputs by implementing Radford M. Neal's xsum
algorithm using the large superaccumulator[1].
The large superaccumulator is used when the input is an array and its size exceeds
PRECISE_SUM_THRESHOLD (1,000); otherwise, the small superaccumulator is used.
The original implementation of xsum large superaccumulator is in C [2]; a C++ translation
by Keita Nonaka was referenced for integration [3].
This patch also refactors the previous implementation[4] of xsum using the small
superaccumulator to align with the xsum large superaccumulator design.

                                TipOfTree              Patched                Ratio
math-sumPrecise-10              2.6139+-0.1125         2.6545+-0.1376         0.0152x slower
math-sumPrecise-100             17.9892+-0.0930        18.0620+-0.1747        0.0040x slower
math-sumPrecise-1000            172.3968+-1.8629       155.9089+-1.3315       1.1057x faster
math-sumPrecise-10000           1709.5876+-4.2845      1530.4218+-5.2034      1.1170x faster

[1]: https://arxiv.org/abs/1505.05571
[2]: https://gitlab.com/radfordneal/xsum
[3]: https://github.com/Gumichocopengin8/xsum.cpp
[4]: https://github.com/WebKit/WebKit/commit/2682d2ab17a40851f7e6bcff35126e810b0646fd

* JSTests/stress/math-sum-precise.js:
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/PreciseSum.cpp:
(WTF::Xsum::SmallAccumulator::SmallAccumulator):
(WTF::Xsum::SmallAccumulator::addInfNan):
(WTF::Xsum::SmallAccumulator::carryPropagate):
(WTF::Xsum::SmallAccumulator::add1NoCarry):
(WTF::Xsum::SmallAccumulator::incrementWhenValueAdded):
(WTF::Xsum::LargeAccumulator::LargeAccumulator):
(WTF::Xsum::LargeAccumulator::addLchunkToSmall):
(WTF::Xsum::LargeAccumulator::largeAddValueInfNan):
(WTF::Xsum::LargeAccumulator::transferToSmall):
(WTF::Xsum::XsumSmall::XsumSmall):
(WTF::Xsum::XsumSmall::addList):
(WTF::Xsum::XsumSmall::add):
(WTF::Xsum::XsumSmall::compute):
(WTF::Xsum::XsumLarge::XsumLarge):
(WTF::Xsum::XsumLarge::addList):
(WTF::Xsum::XsumLarge::add):
(WTF::Xsum::XsumLarge::compute):
(WTF::PreciseSum::PreciseSum): Deleted.
(WTF::PreciseSum::addList): Deleted.
(WTF::PreciseSum::add): Deleted.
(WTF::PreciseSum::compute): Deleted.
(WTF::PreciseSum::xsumSmallAddInfNan): Deleted.
(WTF::PreciseSum::xsumAdd1NoCarry): Deleted.
(WTF::PreciseSum::xsumCarryPropagate): Deleted.
(WTF::PreciseSum::incrementWhenValueAdded): Deleted.
* Source/WTF/wtf/PreciseSum.h:

Canonical link: https://commits.webkit.org/296157@main
